src/Interpreters/wipePasswordFromQuery.cpp

```diff
@@ -0,0 +1,538 @@
+#include <Parsers/Access/ASTCreateUserQuery.h>
+#include <TableFunctions/TableFunctionFactory.h>
+#include <Common/typeid_cast.h>
+
+
+namespace DB
+{
+
+namespace
+{
+    template <bool check_only>
+    class PasswordWipingVisitor
+    {
+    public:
+        struct Data
+        {
+            bool can_contain_password = false;
+            bool is_create_table_query = false;
+            bool is_create_database_query = false;
+            bool is_create_dictionary_query = false;
+            ContextPtr context;
+        };
+
+        using Visitor = std::conditional_t<
+            check_only,
+            ConstInDepthNodeVisitor<PasswordWipingVisitor, /* top_to_bottom= */ true, /* need_child_accept_data= */ true>,
+            InDepthNodeVisitor<PasswordWipingVisitor, /* top_to_bottom= */ true, /* need_child_accept_data= */ true>>;
+
+        static bool needChildVisit(const ASTPtr & /* ast */, const ASTPtr & /* child */, Data & data)
+        {
+            if constexpr (check_only)
+            {
+                return !data.can_contain_password;
+            }
+            else
+            {
+                return true;
+            }
+        }
+
+        static void visit(ASTPtr ast, Data & data)
+        {
+            if (auto * create_user_query = ast->as<ASTCreateUserQuery>())
+            {
+                visitCreateUserQuery(*create_user_query, data);
+            }
+            else if (auto * create_query = ast->as<ASTCreateQuery>())
+            {
+                visitCreateQuery(*create_query, data);
+            }
+            else if (auto * backup_query = ast->as<ASTBackupQuery>())
+            {
+                visitBackupQuery(*backup_query, data);
+            }
+            else if (auto * storage = ast->as<ASTStorage>())
+            {
+                if (data.is_create_table_query)
+                    visitTableEngine(*storage, data);
+                else if (data.is_create_database_query)
+                    visitDatabaseEngine(*storage, data);
+            }
+            else if (auto * dictionary = ast->as<ASTDictionary>())
+            {
+                if (data.is_create_dictionary_query)
+                    visitDictionaryDef(*dictionary, data);
+            }
+            else if (auto * function = ast->as<ASTFunction>())
+            {
+                visitFunction(*function, data);
+            }
+        }
+
+    private:
+        static void visitCreateUserQuery(ASTCreateUserQuery & query, Data & data)
+        {
+            if constexpr (check_only)
+            {
+                data.can_contain_password = true;
+                return;
+            }
+            query.show_password = false;
+        }
+
+        static void visitCreateQuery(ASTCreateQuery & query, Data & data)
+        {
+            if (query.is_dictionary)
+                data.is_create_dictionary_query = true;
+            else if (query.table)
+                data.is_create_table_query = true;
+            else
```

Comment:
I think its better to check `else if (query.database)` if we want to check for database query. As else could also be other things like view, etc. 
Also should masking be handled for view too ?

Reply:
Views don't have anything which needs to be hidden in logs